### DISCUSSION TITLE
Question about the Primehub.imu.heading() and Primehub.imu.reset_heading()

### DISCUSSION THREAD

**Discussion by Digital-Dominators** [pybricks/support#1767] [Q&A]
Hi,
We are new to Pybricks (Our FLL team used SpikePrime block coding last season and this is our 2nd season). 

I was testing the imu.heading() and imu.reset_heading() method, and I did the following:

```python
hub = PrimeHub()
left_motor=Motor(Port.C, Direction.COUNTERCLOCKWISE)
right_motor=Motor(Port.B)

drive_base=DriveBase(left_motor, right_motor, wheel_diameter=56,axle_track=112)
drive_base.use_gyro(True)

print(f'Before test: heading is {hub.imu.heading()}')

print('1st turn 60 degrees--------------')
drive_base.turn(60)
print(f'After 1st turn: heading is {hub.imu.heading()}')

print('Reset heading -------------------')
hub.imu.reset_heading(0)
print(f'After reset: heading is {hub.imu.heading()}')

print('2nd turn 60 degrees--------------')
drive_base.turn(60)
print(f'After 2nd turn: heading is {hub.imu.heading()}')
````

I made 2 60-degree turns and reset the heading in the middle.  The imu.heading() was reset to 0 before the 2nd turn. But after the 2nd 60 dgree turn, it became 120 (which I expect to be 60).  I wonder if it is a bug (which isn't likely as it is a very common method), or it is me who misunderstood these 2 functions  imu.heading()  and imu.reset_heading().  

Could someone help me here and let me know how to code the 2nd turn to get the heading() reading to 60 after the turn? 

This is the print out

```
Before test: heading is 0.0
1st turn 60 degrees--------------
After 1st turn: heading is 65.1416
Reset heading -------------------
After reset: heading is 0.0
2nd turn 60 degrees--------------
After 2nd turn: heading is 118.3906
```
(Also I noticed, the robot turned and accelerated much faster in the 2nd turn, though I didn't change any turn_rate or acceleration parameters in the middle),

**Comment by laurensvalk** [comment]
Thanks for reporting! We will look into it.

**Reply by laurensvalk** [reply]
Sorry for the delay on this. The issue is the following:

```python
drive_base.turn(60) # After this, the robot is actively holding the 60-degree position.

hub.imu.reset_heading(0) # But this tells the current position to be zero, so the 60 degree position is 60 degrees away again!

# It was still trying to hold 60-degrees, so it is going to rapidly try to self-correct, thus appearing to suddenly accelerate quickly.
```

We are currently revising and improving these methods. So my question for you is:
- What would you like to do?
- What did you expect to happen?

This will help us add the right functionality, as well as better document the existing functions like these. Thank you!

**Reply by zoligyenge** [reply]
Trying to understand the relationship between drivebase angle and heading

So if I get it right: Whenever I want to e.g. align to a wall, I would need to reset both heading and drivebase angle, right?

> Sorry for the delay on this. The issue is the following:
> 
> ```python
> drive_base.turn(60) # After this, the robot is actively holding the 60-degree position.
> 
> hub.imu.reset_heading(0) # But this tells the current position to be zero, so the 60 degree position is 60 degrees away again!
> 
> # It was still trying to hold 60-degrees, so it is going to rapidly try to self-correct, thus appearing to suddenly accelerate quickly.
> ```
> 
> We are currently revising and improving these methods. So my question for you is:
> 
> * What would you like to do?
> * What did you expect to happen?
> 
> This will help us add the right functionality, as well as better document the existing functions like these. Thank you!

**Reply by laurensvalk** [reply]
In the latest beta version we released, this has been improved. A more precise definition is given in the documentation that comes with it. 

In the new version, if you have a drivebase that is using the gyro and you reset the drivebase  angle, the gyro will be set to the same value.

In other words, you can do everything through the drivebase now, and don't have to poke at the imu anymore.

This should be more intuitive.

The drivebase reset method now also stops the motor from holding, which is also what you want in that case. This way it won't try to re-correct after aligning with the wall, as it did when this issue was originally opened.

**Reply by zoligyenge** [reply]
Cool. Thanks
